docs/ql-libraries/dataflow/dataflow.md

```diff
@@ -0,0 +1,472 @@
+takes care of matching call-sites with returns and field writes with reads to
+ensure that the generated paths are well-formed. The library also supports a
+number of additional features for improving precision, for example pruning
+infeasible paths based on type information.
+
+## File organisation
+
+The data-flow library consists of a number of files typically located in
+`<lang>/dataflow` and `<lang>/dataflow/internal`:
+
+```
+dataflow/DataFlow.qll
+dataflow/internal/DataFlowImpl.qll
+dataflow/internal/DataFlowCommon.qll
+dataflow/internal/DataFlowImplSpecific.qll
+```
+
+`DataFlow.qll` provides the user interface for the library and consists of just
+a few lines of code importing the implementation:
+
+#### `DataFlow.qll`
+```ql
+import <lang>
+
+module DataFlow {
+  import semmle.code.java.dataflow.internal.DataFlowImpl
+}
+```
+
+The `DataFlowImpl.qll` and `DataFlowCommon.qll` files contain the library code
+that is shared across languages. These contain `Configuration`-specific and
+`Configuration`-independent code, respectively. This organization allows
+multiple copies of the library (for the use case when a query wants to use two
+instances of global data flow and the configuration of one depends on the
+results from the other). Using multiple copies just means duplicating
+`DataFlow.qll` and `DataFlowImpl.qll`, for example as:
+
+```
+dataflow/DataFlow2.qll
+dataflow/DataFlow3.qll
+dataflow/internal/DataFlowImpl2.qll
+dataflow/internal/DataFlowImpl3.qll
+```
+
+The `DataFlowImplSpecific.qll` provides all the language-specific classes and
+predicates that the library needs as input and is the topic of the rest of this
+document.
+
+This file must provide two modules named `Public` and `Private`, which the
+shared library code will import publicly and privately, respectively, thus
+allowing the language-specific part to choose which classes and predicates
+should be exposed by `DataFlow.qll`.
+
+A typical implementation looks as follows, thereby organizing the predicates in
+two files, which we'll subsequently assume:
+
+#### `DataFlowImplSpecific.qll`
+```ql
+module Private {
+  import DataFlowPrivate
+}
+
+module Public {
+  import DataFlowPublic
+}
+```
+
+## Defining the data-flow graph
+
+The main input to the library is the data-flow graph. One must define a class
+`Node` and an edge relation `simpleLocalFlowStep(Node node1, Node node2)`. The
+`Node` class should be in `DataFlowPublic`.
+
+Recommendations:
+* Make `Node` an IPA type. There is commonly a need for defining various
+  data-flow nodes that are not necessarily represented in the AST of the
+  language.
+* Define `predicate localFlowStep(Node node1, Node node2)` as an alias of
+  `simpleLocalFlowStep` and expose it publicly. The reason for this indirection
+  is that it gives the option of exposing local flow augmented with field flow.
+  See the C/C++ implementation, which makes use of this feature.
```

Comment:
Could also mention that it allows for removing (synthesized) local steps that are only relevant for global flow, which [C# does](https://github.com/github/codeql/blob/795c5784b0523d1e6914ab7e592f2abeaece1722/csharp/ql/src/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll#L582-L611).